internal/v4/common_test.go

```diff
@@ -194,17 +195,85 @@ func (s *commonSuite) addPublicCharmFromRepo(c *gc.C, charmName string, rurl *ro
 }
 
 func (s *commonSuite) addPublicCharm(c *gc.C, ch charm.Charm, rurl *router.ResolvedURL) (*router.ResolvedURL, charm.Charm) {
+	return s.addCharm(c, publicCharm(ch), rurl)
+}
+
+type testCharm struct {
+	charm.Charm
+	channels []mongodoc.Channel
+	acls     map[mongodoc.Channel]mongodoc.ACL
+}
+
+func publicCharm(ch charm.Charm) testCharm {
+	return charmOnChannel(charmWithACL(ch, "stable.read", params.Everyone), mongodoc.StableChannel)
+}
+
+func charmWithACL(ch charm.Charm, which string, acl ...string) testCharm {
+	tc, ok := ch.(testCharm)
+	if !ok {
+		tc.Charm = ch
+	}
+	if tc.acls == nil {
+		tc.acls = make(map[mongodoc.Channel]mongodoc.ACL)
+	}
+	parts := strings.SplitN(which, ".", 2)
+	a := tc.acls[mongodoc.Channel(parts[0])]
+	switch parts[1] {
+	case "read":
+		a.Read = acl
+	case "write":
+		a.Write = acl
+	}
+	tc.acls[mongodoc.Channel(parts[0])] = a
+	return tc
+}
+
+func charmOnChannel(ch charm.Charm, channels ...mongodoc.Channel) testCharm {
+	tc, ok := ch.(testCharm)
+	if !ok {
+		tc.Charm = ch
+	}
+	for _, c := range channels {
+		if c != mongodoc.NoChannel && c != mongodoc.UnpublishedChannel {
+			tc.channels = append(tc.channels, c)
+		}
+	}
+	return tc
+}
+
+func (s *commonSuite) addCharm(c *gc.C, ch charm.Charm, rurl *router.ResolvedURL) (*router.ResolvedURL, charm.Charm) {
+	if ch == nil {
+		ch = storetesting.NewCharm(nil)
+	}
+	if tch, ok := ch.(testCharm); ok {
+		rurl, ch = s.addCharm(c, tch.Charm, rurl)
```

Comment:
This recursive call seems a bit squiggly to me. It seems odd that we're passing in a non-nil charm.Charm where if you call any charm.Charm method it'll panic that somehow turns into a valid charm.Charm along the way...

Perhaps it might be nicer to do declare an interface that can optionally be implemented to add on-publish perms and channels; something like this
perhaps?
	// PublishEntity is the type recognized by addCharm and addBundle
	// as a way of associating ACL and channel information with a
	// charm at publish time.
	type PublishEntity interface {
		ACLs() map[mongodoc.Channel]mongodoc.ACL
		Channels() []mongodoc.Channel
	}
	
	// ... probably use this type in Store.AddCharmWithArchive and avoid
	// the current type switch shenanigans (which are wrongly documented
	// anyway)
	type CharmWithArchive interface {
		charm.Charm
		charmstore.ArchiverTo
	}
	
	
	type charmWithPublishInfo struct {
		charmstore.CharmWithArchive
		PublishEntity
	}
	
	func charmWithACLs(ch CharmWithArchive, acls map[mongodoc.Channel]mongodoc.ACL) CharmWithArchive {
		return charmWithPublishInfo{
			PublishEntity: withACLs(ch, acls),
			CharmWithArchive: ch,
		}
	}
	
	func charmWithChannels(ch CharmWithArchive, channels ...mongodoc.Channel) CharmWithArchive {
		return charmWithPublishInfo{
			PublishEntity: withChannels(ch, acls),
			CharmWithArchive: ch,
		}
	}
	
	type bundleWithPublishInfo struct {
		charmstore.BundleWithArchive
		PublishEntity
	}
		
	func bundleWithACLs(b BundleWithArchive, acls map[mongodoc.Channel]mongodoc.ACL) CharmWithArchive {
		return bundleWithPublishInfo{
			PublishEntity: withACLs(ch, acls),
			BundleWithArchive: b,
		}
	}
	
	func bundleWithChannels(b BundleWithArchive, channels ...mongodoc.Channel) CharmWithArchive {
	
	
	func withACLs(x interface{}, acls map[mongodoc.Channel]mongodoc.ACL) PublishEntity {
		info := publishEntityInfo{
			acls: acls,
		}
		if x, ok  := x.(PublishEntity); ok {
			info.channels = x.Channels()
		}
		return info
	}
	
	func withChannels(x interface{}, channels ...mongodoc.Channel) PublishEntity {
		info := publishEntityInfo{
			channels: channels,
		}
		if x, ok  := x.(PublishEntity); ok {
			info.acls = x.ACLs()
		}
		return info
	}
	
	type publishEntityInfo struct {
		acls     map[mongodoc.Channel]mongodoc.ACL
		channels []mongodoc.Channel
	}
	
	func (a publishEntityInfo) ACLs() map[mongodoc.Channel]mongodoc.ACL {
		return a.acls
	}
	
	func (a publishEntityInfo) Channels() []mongodoc.Channel {
		return a.channels
	}